xclim/subset.py

```diff
@@ -125,11 +133,188 @@ def func_checker(*args, **kwargs):
+        except (KeyError, ValueError):
+            logging.exception("No shape provided.")
+            raise
+
+        if "use_all_features" in kwargs:
+            use_all_features = bool(kwargs["use_all_features"])
+        else:
+            use_all_features = False
+
+        geom = list()
+        geometry_types = list()
+        try:
+            with fiona.open(shape) as fio:
+                logging.info("Vector read OK.")
+                if "crs" not in kwargs:
+                    kwargs["shape_crs"] = rasterio.crs.CRS(fio.crs)
+
+                for i, feat in enumerate(fio):
+                    g = geojson.GeoJSON(feat)
+                    geom.append(g["geometry"])
+                    geometry_types.append(g["geometry"]["type"])
+                    if not use_all_features:
+                        break
+        except fiona.errors.DriverError:
+            logging.exception("Unable to read shape.")
+            raise
+
+        if len(geom):
+            kwargs["geometry"] = geom
+            logging.info("Shapes found are {}.".format(", ".join(set(geometry_types))))
+            return func(*args, **kwargs)
+        else:
+            raise RuntimeError("No geometries found.")
+
+    return func_checker
+
+
+@check_geometry_and_read
+@check_date_signature
+def subset_shape(
+    da: Union[xarray.DataArray, xarray.Dataset],
+    shape: Union[str, Path],
+    start_date: Optional[str] = None,
+    end_date: Optional[str] = None,
+    geometry: Optional[List[geojson.GeoJSON]] = None,
+    da_crs: Optional[str] = None,
+    shape_crs: Optional[str] = None,
+    use_all_features: bool = False,
+) -> Union[xarray.DataArray, xarray.Dataset]:
+    """Subset a DataArray or Dataset spatially (and temporally) using a vector shape and date selection.
+
+    Return a subsetted data array for grid points falling within the area of a polygon and/or MultiPolygon shape,
+      or grid points along the path of a LineString and/or MultiLineString.
+
+    Parameters
+    ----------
+    da : Union[xarray.DataArray, xarray.Dataset]
+      Input data.
+    shape : Union[str, Path, geometry.GeometryCollection]
+      Path to a single-layer vector file, or a shapely GeometryCollection object.
+    start_date : Optional[str]
+      Start date of the subset.
+      Date string format -- can be year ("%Y"), year-month ("%Y-%m") or year-month-day("%Y-%m-%d").
+      Defaults to first day of input data-array.
+    end_date : Optional[str]
+      End date of the subset.
+      Date string format -- can be year ("%Y"), year-month ("%Y-%m") or year-month-day("%Y-%m-%d").
+      Defaults to last day of input data-array.
+    da_crs : Optional[Union[int, dict, str]]
+      CRS of the xarray.DataArray or xarray.Dataset provided. Default: dict(epsg=4326).
+    shape_crs : Optional[Union[int, dict, str]]
+      CRS of the geometries provided. If passing GeometryCollections as shapes, CRS must be explicitly stated.
+    use_all_features : bool
```

Comment:
That seems counter-intuitive ? Why would you pass a shapefile and not want to use all features ?

Reply:
Also, this argument is for the benefit only of the decorator. I don't think this is the intended use of decorators.

Reply:
I added this in order to quickly test for shapefiles and left it in as a "feature". Removing in now.